### PR TITLE
M2m prototype

### PR DESCRIPTION
**Pull request by ma-gu**
@johnkattenhorn : please have a look on this. I hope this approach make some sense for you.

### CODE REVIEW

File: WorkerActorService/ProcessorActor.cs [hunk 1]
```diff
@@ -22,14 +22,21 @@ namespace Microsoft.AzureCat.Samples.WorkerActorService
...
     /// </remarks>
     [ActorService(Name = "ProcessorActorService")]
     [StatePersistence(StatePersistence.Persisted)]
     internal class ProcessorActor : Actor, IProcessorActor
     {
+        private IMessageProcessorDispatcher Dispathcer { get; }
+
+        public ProcessorActor(IMessageProcessorDispatcher dispatcher)
+        {
+            Dispathcer = dispatcher;
```
Comment (aprooks) [inline WorkerActorService/ProcessorActor.cs]: `Dispathcer` typo

**Comment by aprooks** [comment]
looks good. Few points
1. Using IoC container feels like an overkill. You probably won't have too many dependencies to configure or switch: couple of connection strings, logger settings, anything else? Most probably static container would be enough. 
1. Inbound/Outbound types are too generic and confusing. E.g. a single message can be published by Device actor and then consumed by another Actor. What would this message be? It might be worth to use semantics of `IEvent` and `ICommand`. Where `IEvent` is a fact of what happened, and `ICommand` is what you send to something and that can be potentially rejected.

File: WorkerActorService/ProcessorActor.cs (hunk 1 above)
Comment (ma-gu) [inline WorkerActorService/ProcessorActor.cs]: @aprooks Hard to notice, I even was considering to use diff tool for it :) 
Fixed.

**Comment by ma-gu** [comment]
@aprooks: 

> Using IoC container feels like an overkill.

Well, everything could be done manually, I agree. One more point - this is a prototype only, in a real project (MKopa.M2M v2 ) configuring dependencies can easily grow up to more complex. But using IoC provide conventional registering for us: everything located it `.Services` namespace and marked with `[assembly: IoC]` attribute will be registered automatically. We have 10 device types in MKopa, so we can choose: 

1. Configure dependencies for each device-specific project manually (10 times...);
2. Simply place our implementations to `Services` folder/Namespace and mark our assembly with `IoC`.

To my taste the 2-nd option looks like much less typing.

>Inbound/Outbound types are too generic and confusing.

Good point. I just used terminology/message types from real project. In terms of generality your suggestion sounds good. 
@johnkattenhorn: How do you think, does such type of Actor-Actor message processing make sense for M2M project?

**Comment by johnkattenhorn** [comment]
I think @ma-gu you are right, the number of Devices type we have for V1 is now 10 and V2 is already at 4 I think, especially with the V2 devices they are becoming very diverse. Although we probably didn't need IoC for the PoC we need to make sure that such implementation detail will work in production so it's better to find out now. The time we should need to consider not using IoC would be if retrieving from the container takes too much time I guess. @jkonecki already has some ideas to share on optimising this I think.

On the `ICommand`, `IEvent` pattern, I think this is pretty good from an outbound perspective as we have many different actions that the device actor has to take on, on the inbound I'm not sure how relevant it is given that it is essentially one command 'process this please', maybe you could argue that this is just another type of command ? @jkonecki - What's you thoughts ?

**Comment by johnkattenhorn** [comment]
Good to have the input from everyone, if I'm honest I'm getting nervous as I need to get this PoC producing some realistic performance metric with using some IBIS1 workload asap as M-Kopa want a revised delivery plan for next week.

**Comment by aprooks** [comment]
> Configure dependencies for each device-specific project manually (10 times...);

**Pull request merged**